File: android/src/main/java/com/example/flutter_file_downloader/core/DownloadTaskBuilder.java
package com.example.flutter_file_downloader.core;

import android.app.Activity;

public class DownloadTaskBuilder {
//    DownloadTaskBuilder instance;
    String url, name;
    DownloadCallbacks callbacks;
    boolean deleteOldVersion = false;
    final Activity activity;
    private DownloadTask task;

    public DownloadTaskBuilder(final Activity activity) {
        this.activity = activity;
//        instance = this;
    }

    public DownloadTaskBuilder setUrl(final String url) {
        this.url = url;
        return this;
    }

    public DownloadTaskBuilder setName(final String name) {
        this.name = name;
        return this;
    }

    public DownloadTaskBuilder setCallbacks(final DownloadCallbacks callbacks) {
        this.callbacks = callbacks;
        return this;
    }

    public DownloadTaskBuilder setDeleteOldVersion(final boolean delete) {
        deleteOldVersion = delete;
        return this;
    }

    public DownloadTask build() {
        if (task == null)
            task = new DownloadTask(
                    activity,
                    url,
                    name,
                    deleteOldVersion,
                    callbacks);
        try {
            return getDownloadTask();
        } catch (Exception e) {
            return task;
        }
    }

    public DownloadTask getDownloadTask() throws Exception {
        if (task == null)
            throw new Exception("build method is not called, you should call \"downloadTaskBuilder.build()\" first");
        return task;
    }
}

File: ios/Classes/core/DownloadTaskBuilder.swift
import Foundation
import UIKit

class DownloadTaskBuilder {
    private var instance: DownloadTaskBuilder
    private var url: String?
    private var name: String?
    private var callbacks: DownloadCallbacks?
    private let activity: Activity
    private var task: DownloadTask?

    init(activity: Activity) {
        self.activity = activity
        instance = self
    }

    func setUrl(url: String) -> DownloadTaskBuilder {
        instance.url = url
        return instance
    }

    func setName(name: String) -> DownloadTaskBuilder {
        instance.name = name
        return instance
    }

    func setCallbacks(callbacks: DownloadCallbacks) -> DownloadTaskBuilder {
        instance.callbacks = callbacks
        return instance
    }

    func build() -> DownloadTask {
        if task == nil {
            task = DownloadTask(activity: instance.activity, url: instance.url!, name: instance.name, callbacks: callbacks)
        }
        return getDownloadTask()
    }

    func getDownloadTask() -> DownloadTask {
        if task == nil {
            fatalError("build method is not called, you should call \"downloadTaskBuilder.build()\" first")
        }
        return task!
    }
}
